rayon-demo/src/sort.rs

```diff
@@ -49,32 +51,69 @@ fn gen_strings(len: usize) -> Vec<String> {
 }
 
 fn gen_big_random(len: usize) -> Vec<[u64; 16]> {
-    let mut rng = thread_rng();
+    let mut rng = XorShiftRng::from_seed([0, 1, 2, 3]);
     rng.gen_iter().map(|x| [x; 16]).take(len).collect()
 }
 
 macro_rules! sort {
     ($f:ident, $name:ident, $gen:expr, $len:expr) => {
         #[bench]
         fn $name(b: &mut Bencher) {
-            b.iter(|| $gen($len).$f());
+            let v = $gen($len);
+            b.iter(|| v.clone().$f());
             b.bytes = $len * mem::size_of_val(&$gen(1)[0]) as u64;
         }
     }
 }
 
-sort!(par_sort, par_sort_large_ascending, gen_ascending, 50_000);
-sort!(par_sort, par_sort_large_descending, gen_descending, 50_000);
-sort!(par_sort, par_sort_large_mostly_ascending, gen_mostly_ascending, 50_000);
-sort!(par_sort, par_sort_large_mostly_descending, gen_mostly_descending, 50_000);
-sort!(par_sort, par_sort_large_random, gen_random, 50_000);
-sort!(par_sort, par_sort_large_big_random, gen_big_random, 50_000);
-sort!(par_sort, par_sort_large_strings, gen_strings, 50_000);
+macro_rules! sort_strings {
+    ($f:ident, $name:ident, $gen:expr, $len:expr) => {
+        #[bench]
+        fn $name(b: &mut Bencher) {
+            let v = $gen($len);
+            let v = v.iter().map(|s| &**s).collect::<Vec<&str>>();
+            b.iter(|| v.clone().$f());
+            b.bytes = $len * mem::size_of::<&str>() as u64;
+        }
+    }
+}
+
+macro_rules! sort_expensive {
+    ($f:ident, $name:ident, $gen:expr, $len:expr) => {
+        #[bench]
+        fn $name(b: &mut Bencher) {
+            let v = $gen($len);
+            b.iter(|| {
+                let count = AtomicUsize::new(0);
+                let mut v = v.clone();
+                v.$f(|a: &u64, b: &u64| {
+                    count.fetch_add(1, SeqCst);
+                    if count.load(SeqCst) % 1_000_000_000 == 0 {
+                        panic!("should not happen");
```

Comment:
Is this making sure we don't hit O(n<sup>2</sup>) comparisons? (since 50,000<sup>2</sup> > 10<sup>9</sup>)  A comment would be nice.

The `fetch_add` followed by `load` is strange too, but doesn't matter much since this is just toy benchmark code...